### PR TITLE
Document FueledUtils

### PR DESCRIPTION
**Pull request by stephanecopin**
This PR adds a lot of missing documentation to `FueledUtils`.
It also deprecates some stuff, and update the public interface of others, all while retaining backward compatibility so nothing breaks. I'll be justifying some of my decisions by commenting on the PR.

Sometimes I went the other route, and decided _not to_ change the public interface so as to not break compatibility, but I do wish to update it in another PR.

The documentation files is generated via `Jazzy`, using `jazzy`

@leontiy I've not documented methods such as `transitionContext` or `transitionContext`, as I've never used them myself and I'm unsure how to use them properly (which is why they probably should documented). When/if you have some free time, it'd be great to have them documented

### CODE REVIEW

File: FueledUtils/DecoratingTextFieldDelegate.swift [hunk 1]
```diff
@@ -73,44 +90,56 @@ public final class DecoratingTextFieldDelegate: NSObject {
...
-	///		- dataString: a string conaining only data characters (see `isDataCharacter`).
-	/// - Returns: a representation of `dataString` formatted using a corresponding pattern obtained using \
-	/// `patternForDataString`.
+	///		- dataString: A string containing only data characters (see `isDataCharacter`).
+	///
+	/// - Returns: The converted string from the input.
+	///
 	public func decorateString(_ dataString: String) -> String {
-		var res = ""
+		var result = ""
```
Comment (stephanecopin) [inline FueledUtils/DecoratingTextFieldDelegate.swift:102]: I've done this a few times in the code, to avoid abbreviations. Sometimes it got pretty absurd with abbreviations such as `bw`, `cx` or `nc`, which [obfuscated the code IMHO](https://github.com/Droogans/unmaintainable-code#cd-wrttn-wtht-vwls-s-mch-trsr). These names just forced the user to read where the variable was declared to find out what it referred too, breaking the natural reading flow of the code.

File: FueledUtils/DecoratingTextFieldDelegate.swift [hunk 2]
```diff
@@ -73,44 +90,56 @@ public final class DecoratingTextFieldDelegate: NSObject {
...
-		return res
+		return result
 	}
 
-	fileprivate func convertDecoRange(_ decoRange: NSRange, fromDecoratedString decoratedString: String) -> NSRange {
-		let decoPrefix = (decoratedString as NSString).substring(to: decoRange.location)
-		let decoSubstring = (decoratedString as NSString).substring(with: decoRange)
-		let dataPrefix = self.undecorateString(decoPrefix)
-		let dataSubstring = self.undecorateString(decoSubstring)
+	fileprivate func convertDecoratedRange(_ decoratedRange: NSRange, fromDecoratedString decoratedString: String) -> NSRange {
```
Comment (stephanecopin) [inline FueledUtils/DecoratingTextFieldDelegate.swift:138]: I made the code consistent by spelling `decorated` everywhere

File: FueledUtils/DimmingButton.swift [hunk 3]
```diff
@@ -16,25 +16,43 @@ limitations under the License.
...
+			self.updateDimmedAmount(for: self.dimmingView, dimAmount: 1.0)
+		}
+		didSet {
+			self.updateDimmedViewAmount()
+		}
+	}
+	///
+	/// The alpha to set the view to when dimmed. Defaults to `0.4`
+	///
+	@IBInspectable public var dimAmount: CGFloat = 0.4 {
```
Comment (stephanecopin) [inline FueledUtils/DimmingButton.swift]: Maybe `dimmingAmount` here?

File: FueledUtils/RACExtensions.swift [hunk 4]
```diff
@@ -320,6 +338,7 @@ extension Reactive where Base: UIViewController {
 
 @available(iOS 9.0, *)
 extension Reactive where Base: UIStackView {
+	@available(*, deprecated, message: "Use `subview.reactive.isHidden <~ <Binding Source>` instead")
 	public func isArranged(_ subview: UIView, at index: Int) -> BindingTarget<Bool> {
```
Comment (stephanecopin) [inline FueledUtils/RACExtensions.swift:342]: When a view is in a `UIStackView`, hiding it will update the constraints so it's as if the view was removed.

File: FueledUtils/ScrollViewPage.swift [hunk 5]
```diff
@@ -16,24 +16,71 @@ limitations under the License.
...
+	}
+
+	///
+	/// Gets/Sets (without animation) the current page of the scroll view, assuming a paginated scroll view of width `self.bounds.size.width`, with no left or right content insets.
+	///
+	/// To set the current page with an animation, use `setCurrentPage(_ page:, animated:)`
+	///
+	/// - Returns: Returns the current page (0-based)
+	///
+	@nonobjc public var currentPage: Int {
```
Comment (stephanecopin) [inline FueledUtils/ScrollViewPage.swift]: If `@nonobjc` is specified, it can't conflict with a potential property of the same name in Objective-C, making prefix reliable. Granted it could still conflict if the same property is defined in different module, so let me know what you think

File: FueledUtils/StringExtensions.swift [hunk 6]
```diff
@@ -15,31 +15,45 @@ limitations under the License.
...
-	}
-
+extension StringProtocol {
 	public var nsLength: Int {
-		return (self as NSString).length
+		return (String(self) as NSString).length
 	}
 
+	@available(*, deprecated, renamed: "nsRange")
 	public var fullRange: NSRange {
```
Comment (stephanecopin) [inline FueledUtils/StringExtensions.swift:24]: For consistency with `nsLength`, I've renamed this one `nsRange`

File: FueledUtils/UIExtensions.swift [hunk 7]
```diff
@@ -114,45 +150,124 @@ public extension UIView {
...
+		}
+	}
+
+	///
+	/// **Deprecated**: Please use `deselectAllRows(animated:)` instead.
+	///
+	/// Refer to the documentation for `deselectAllRows(animated:)` for more info.
+	///
+	@available(*, deprecated, renamed: "deselectAllRows(animated:)")
 	public func deselectAllRows(_ animated: Bool) {
```
Comment (stephanecopin) [inline FueledUtils/UIExtensions.swift:189]: I've named the parameter `animated` for consistency with other functions, as `deselectAllRows(true)` doesn't make sense without reading its signature

File: FueledUtils/UIExtensions.swift [hunk 8]
```diff
@@ -114,45 +150,124 @@ public extension UIView {
...
-		var image: UIImage?
+	///
+	/// Create a `CGContext` allowing to do custom drawing, and returns the resulting image as drawn in the context.
+	/// - Parameters:
+	///   - size: Please refer to the parameters documentation for `UIGraphicsBeginImageContextWithOptions` for more info.
+	///   - siopaqueze: Please refer to the parameters documentation for `UIGraphicsBeginImageContextWithOptions` for more info.
+	///   - scale: Please refer to the parameters documentation for `UIGraphicsBeginImageContextWithOptions` for more info.
+	///   - graphics: The drawing actions to execute.
+	///
+	public static func draw(size: CGSize, opaque: Bool = false, scale: CGFloat = 0.0, graphics: (CGContext) -> Void) -> UIImage {
```
Comment (stephanecopin) [inline FueledUtils/UIExtensions.swift:252]: I want to update this method later on to return the `optional` it should return, for example if `size`'s `width` or `height` is 0.0, the resulting image will be `nil`. The user should be responsible for force-casting if he's sure the image is valid.

File: FueledUtils/UIExtensions.swift [hunk 9]
```diff
@@ -75,7 +93,17 @@ public extension UILabel {
...
 	public var fueled_animating: Bool {
+		get {
+			return self.animating
+		}
+		set {
+			self.animating = newValue
+		}
+	}
+
+	@nonobjc public var animating: Bool {
```
Comment (stephanecopin) [inline FueledUtils/UIExtensions.swift]: Maybe keep the same name here?

File: FueledUtils/UIExtensions.swift [hunk 10]
```diff
@@ -114,68 +150,190 @@ public extension UIView {
...
+		return image
 	}
 
+	///
+	/// **Deprecated**: Please use `withAlpha(_:)` instead.
+	///
+	/// Refer to the documentation for `withAlpha(_:)` for more info.
+	///
+	@available(*, deprecated, renamed: "withAlpha(_:)")
 	public func imageWithAlpha(_ alpha: CGFloat) -> UIImage {
```
Comment (stephanecopin) [inline FueledUtils/UIExtensions.swift:269]: Made this method and the next few ones more "swifty"

File: FueledUtils/UIExtensions.swift [hunk 11]
```diff
@@ -114,68 +150,190 @@ public extension UIView {
...
+	///   - offset: The offset to apply to the original image in the resulting image.
+	///   - contextSize: The size of the context used to resize the image. If not specified, it defaults to `CGSize(width: offset.x + size.width, height: offset.y + size.height)`
+	///   - size: The size to resize the image to.
+	///
+	public func resized(offset: CGPoint = .zero, contextSize: CGSize? = nil, size: CGSize) -> UIImage? {
+		let contextSize = contextSize ?? CGSize(width: offset.x + size.width, height: offset.y + size.height)
+		return UIImage.draw(size: contextSize) { _ in
 			self.draw(in: CGRect(origin: offset, size: size))
-		}
+		}.withRenderingMode(self.renderingMode)
```
Comment (stephanecopin) [inline FueledUtils/UIExtensions.swift:328]: Breaks backward compatibility, but for consistency with the others I've added it

File: FueledUtils/ScrollViewPage.swift (hunk 5 above)
Comment (leontiy) [reply FueledUtils/ScrollViewPage.swift]: @stephanecopin `@nonobjc` is now default, so no need to specify it. I checked. only adding `@objc` makes the methods show up in the objective-c runtime reflection.

File: FueledUtils/UIExtensions.swift (hunk 7 above)
Comment (leontiy) [reply FueledUtils/UIExtensions.swift:189]: 👍

**Comment by leontiy** [comment]
I'm pretty sure I extensively documented `transitionContext` et al. with usage examples…

**Comment by leontiy** [comment]
@stephanecopin see the docs around `transitionContext` for usage examples. But you are right in that the doc for transitionContext should link to the observe(inContext:) function on signal protocol.

File: FueledUtils/DecoratingTextFieldDelegate.swift (hunk 1 above)
Comment (leontiy) [reply FueledUtils/DecoratingTextFieldDelegate.swift:102]: I've actually recently done the same in local stuff derived from these for vertical pagination.

File: FueledUtils/UIExtensions.swift (hunk 9 above)
Comment (leontiy) [reply FueledUtils/UIExtensions.swift]: should be isAnimating, right?

File: FueledUtils/CollectionExtensions.swift [hunk 12]
```diff
@@ -14,8 +14,14 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 public extension Collection {
+	///
+	/// Try to get the item at index `index`. If the index is out of bounds, `nil` is returned.
+	///
+	/// Parameter index: The index of the item to tentatively get.
+	/// Returns: The element as a wrapped optional if the `index` is in the `indices` of the collection, `nil` otherwise
+	///
 	public func getSafely(_ index: Self.Index) -> Self.Iterator.Element? {
```
Comment (leontiy) [inline FueledUtils/CollectionExtensions.swift]: should be `getSafely(at index: Self.Index)…` IMO

File: FueledUtils/DimmingButton.swift (hunk 3 above)
Comment (leontiy) [reply FueledUtils/DimmingButton.swift]: dimmedAlpha

File: FueledUtils/UIExtensions.swift (hunk 9 above)
Comment (stephanecopin) [reply FueledUtils/UIExtensions.swift]: It should, but `isAnimating` is already taken by `UIActivityIndicatorView` :| The only solution I see would be to create an Objective-C wrapper, but I'm not sure how desirable that is

File: FueledUtils/CollectionExtensions.swift (hunk 12 above)
Comment (stephanecopin) [reply FueledUtils/CollectionExtensions.swift]: Yep I agree, let me make that change, I'll deprecate the old one

File: FueledUtils/KeyboardInsetHelper.swift [hunk 13]
```diff
@@ -63,23 +87,53 @@ open class KeyboardInsetHelper: NSObject {
 		let keyboardFrame = referenceView.convert(keyboardFrameValue.cgRectValue, from: referenceView.window)
 		let curveOption = UIView.AnimationOptions(rawValue: curve << 16)
 		let animationOptions = curveOption.union(.beginFromCurrentState)
-		let inset = max(baseInset, referenceView.bounds.maxY - keyboardFrame.minY)
+		let baseInset = referenceView.bounds.maxY - keyboardFrame.minY
+		let inset = max(minimumInset, referenceView.bounds.maxY - keyboardFrame.minY)
```
Comment (notbenoit) [inline FueledUtils/KeyboardInsetHelper.swift]: Could we use `baseInset ` here in as the `max` second parameter ?

File: FueledUtils/LoadingState.swift [hunk 14]
```diff
@@ -38,6 +58,9 @@ public enum LoadingState<Error: Swift.Error> {
 }
 
 public extension Action {
+	///
+	/// Returns the current loading state for a given action.
+	///
 	public func loadingState() -> SignalProducer<LoadingState<Error>, NoError> {
```
Comment (notbenoit) [inline FueledUtils/LoadingState.swift]: Curious about why we did not make this a computed var instead of function without parameters 🤔 
*EDIT* : but this is not related to documentation anyway 👌

Comment (stephanecopin) [reply FueledUtils/LoadingState.swift]: @notbenoit true, I can easily deprecate it and create a computed property though, Swift is smart enough not to make the code ambiguous in usage. What do you think?

Comment (notbenoit) [reply FueledUtils/LoadingState.swift]: Agreed, let's deprecate it. It'll be more in sync with what `Action` exposes (`isExecuting`, `isEnabled`, `values`, etc...)

File: FueledUtils/ScrollViewPage.swift [hunk 15]
```diff
@@ -16,24 +16,71 @@ limitations under the License.
...
+			let page = Int((self.bounds.size.width / self.contentOffset.x).rounded())
+			return min(max(0, page), self.numberOfPages - 1)
+		}
+		set {
+			self.setCurrentPage(newValue, animated: false)
+		}
 	}
 
+	///
+	/// Gets the current page of the scroll view, assuming a paginated scroll view of width `self.bounds.size.width`, with no left or right content insets.
```
Comment (notbenoit) [inline FueledUtils/ScrollViewPage.swift]: Woops, wrong function documentation 😛

File: FueledUtils/UIExtensions.swift (hunk 10 above)
Comment (notbenoit) [reply FueledUtils/UIExtensions.swift:269]: Could be `with(alpha: CGFloat)` ? Although, the similar `UIColor` function  is `withAlphaComponent(_ alpha: CGFloat)`. We could meet halfway with `withAlpha(_ alpha: CGFloat)`
**EDIT** : `withAlpha(_ alpha: CGFloat)` is what you have 👍 I need glasses I think 👓

**Comment by stephanecopin** [comment]
@leontiy @notbenoit I've implemented/tweaked what you suggested, and marked the relevant comment threads as Resolved. Please do check if it looks good, as I'm thinking this is looking good to merge soon

File: FueledUtils/LoadingState.swift (hunk 14 above)
Comment (heymansmile) [reply FueledUtils/LoadingState.swift]: @stephanecopin I think the documentation should be changed to something like that:
`A producer for Signals that will send the action’s current LoadingState.`

File: FueledUtils/RACExtensions.swift [hunk 16]
```diff
@@ -157,27 +161,41 @@ public extension SignalProtocol {
...
+	///
 	public func ignoreError() -> SignalProducer<Value, NoError> {
 		return self.producer.flatMapError { _ in
 			SignalProducer<Value, NoError>.empty
 		}
 	}
+
+	///
+	/// Starts the `SignalProducer` after the given delay, on the given `DateScheduler`.
+	///
```
Comment (heymansmile) [inline FueledUtils/RACExtensions.swift:175]: @stephanecopin I think this should be changed to something like that:
`Returns a SignalProducer that when started will delay starting of the original producer on given scheduler.`

Comment (stephanecopin) [reply FueledUtils/RACExtensions.swift:175]: Makes sense, updating

File: FueledUtils/LoadingState.swift (hunk 14 above)
Comment (heymansmile) [reply FueledUtils/LoadingState.swift]: @stephanecopin here too ^ :)

File: FueledUtils/LoadingState.swift [hunk 17]
```diff
@@ -38,7 +58,22 @@ public enum LoadingState<Error: Swift.Error> {
...
-	public func loadingState() -> SignalProducer<LoadingState<Error>, NoError> {
+	///
+	/// **Deprecated**: Please use `getSafely(at:)` instead.
+	///
+	/// Refer to the documentation for `getSafely(at:)` for more info.
+	///
+	@available(*, deprecated, renamed: "loadingState")
+	// The unused parameter allows to bypass the compiler error "Invalid redeclaration of 'loadingState'",
+	// while retaining backward compatibility
+	public func loadingState(_ unused: Void = ()) -> SignalProducer<LoadingState<Error>, NoError> {
```
Comment (stephanecopin) [inline FueledUtils/LoadingState.swift:79]: @notbenoit I had to do this unfortunately to make sure the compiler didn't complain, and it seems to work in practice so I'm leaving it as is

File: FueledUtils/LoadingState.swift [hunk 18]
```diff
@@ -38,7 +58,22 @@ public enum LoadingState<Error: Swift.Error> {
...
+		return self.loadingState
+	}
+
+	///
+	/// Returns the current loading state for a given action.
+	///
```
Comment (heymansmile) [inline FueledUtils/LoadingState.swift:86]: @stephanecopin I think the documentation should be changed to something like that:
`A producer for Signals that will send the action’s current LoadingState.`

**Review by heymansmile — CHANGES_REQUESTED** [review]

File: FueledUtils/LoadingState.swift (hunk 18 above)
Comment (stephanecopin) [reply FueledUtils/LoadingState.swift:86]: @heymansmile Makes sense. This actually got me thinking, wouldn't this make more sense to return a `Signal`? For consistency with the rest of the `Action` properties, but also because it doesn't make sense to `start()` this - it's something that is always happening when the action is created. What do you think?
cc @leontiy @notbenoit

Comment (heymansmile) [reply FueledUtils/LoadingState.swift:86]: @stephanecopin No, because since state is a continuous thing we should be able to get this any time we bind loadingState to a property. SignalProducer sent initial values at start in contrast with Signals. So for example when reusing cells with cellModels which contain actions and we want to display current state of the actions we need to bind cell's controls to the signal producers of loading states. Does it makes sense?

File: FueledUtils/LoadingState.swift [hunk 19]
```diff
@@ -38,11 +68,27 @@ public enum LoadingState<Error: Swift.Error> {
...
+	/// **Deprecated**: Please use `getSafely(at:)` instead.
+	///
+	/// Refer to the documentation for `getSafely(at:)` for more info.
+	///
+	@available(*, deprecated, renamed: "loadingState")
+	// The unused parameter allows to bypass the compiler error "Invalid redeclaration of 'loadingState'",
+	// while retaining backward compatibility
+	public func loadingState(_ unused: Void = ()) -> SignalProducer<LoadingState<Error>, NoError> {
+		return self.loadingState.producer
```
Comment (heymansmile) [inline FueledUtils/LoadingState.swift]: This is not the "same" producer as the deprecated one. If a binding occurs in the moment of action's executing the `loading` state will not be sent because this producer will emit the same events as the original signal. (see my comment for the new `var loadingState: Signal<LoadingState<Error>, NoError>` implementation)

**Review by heymansmile — CHANGES_REQUESTED** [review]

File: FueledUtils/LoadingState.swift (hunk 18 above)
Comment (stephanecopin) [reply FueledUtils/LoadingState.swift:86]: @heymansmile Thinking further about it, I think it makes the most sense to make it a `Property` then. What do you think?
A `property` has the semantics you're talking about, while it _always_ has a state and doesn't require starting. We can also get a `producer` or a `signal` from it easily.

Comment (stephanecopin) [reply FueledUtils/LoadingState.swift:86]: I'll also be amending the doc for `.default` above so it reflects exactly when it's sent:
```
Represents the state of an action was never started, has received any values, has completed successfully or was interrupted
```

Comment (heymansmile) [reply FueledUtils/LoadingState.swift:86]: @stephanecopin It would be a kinda breaking change because the initial implementation doesn't send any values if the Action has not been started yet. From my point of view it would be enough for us to replace method with computed property but return a SignalProducer as in the original implementation. A developer then can do everything needed with it depending on a task.

Comment (stephanecopin) [reply FueledUtils/LoadingState.swift:86]: Haha, all right I'll revert it back to what it was 😉

**Review by heymansmile — APPROVED** [review]
